File: ViewModels/CollectionViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using DuelRecords.Scan.Data.Models;
using DuelRecords.Scan.Data.Services;
using DuelRecords.Scan.Pages;

namespace DuelRecords.Scan.ViewModels;

public partial class CollectionViewModel(DuelRecordsApiService apiService) : ObservableObject
{
    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(ShowCards))]
    [NotifyPropertyChangedFor(nameof(ShowEmpty))]
    [NotifyPropertyChangedFor(nameof(ShowError))]
    public partial bool IsLoading { get; set; }

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(ShowCards))]
    [NotifyPropertyChangedFor(nameof(ShowEmpty))]
    public partial List<Card> Cards { get; set; } = [];

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(ShowError))]
    [NotifyPropertyChangedFor(nameof(ShowCards))]
    [NotifyPropertyChangedFor(nameof(ShowEmpty))]
    public partial string? ErrorMessage { get; set; }

    public bool ShowCards => !IsLoading && ErrorMessage is null && Cards.Count > 0;
    public bool ShowEmpty => !IsLoading && ErrorMessage is null && Cards.Count == 0;
    public bool ShowError => !IsLoading && ErrorMessage is not null;

    [RelayCommand]
    private async Task LoadCardsAsync()
    {
        IsLoading = true;
        ErrorMessage = null;
        try
        {
            Cards = await apiService.GetCardsAsync();
        }
        catch (Exception ex)
        {
            ErrorMessage = $"Erro ao carregar coleção: {ex.Message}";
        }
        finally
        {
            IsLoading = false;
        }
    }

    [RelayCommand]
    private static async Task GoToScanAsync()
        => await Shell.Current.GoToAsync(nameof(ScanPage));
}

File: app/src/main/java/com/duelrecords/scan/viewmodel/CollectionViewModel.kt
package com.duelrecords.scan.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.duelrecords.scan.data.api.RetrofitClient
import com.duelrecords.scan.data.model.Card
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class CollectionViewModel : ViewModel() {

    private val _cards = MutableStateFlow<List<Card>>(emptyList())
    val cards: StateFlow<List<Card>> = _cards

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    init {
        loadCards()
    }

    fun loadCards() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                _cards.value = RetrofitClient.api.getCards()
            } catch (e: Exception) {
                _error.value = "Erro ao carregar coleção: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }
}
